### PR TITLE
Fix errors in drag cursors when using high-DPI screens

### PR DESCRIPTION
**Pull request by ptziegler**
The cursors used in the flyout palette only exist for the default 100%. If a system is using a higher zoom level, the creation of those cursors fails with an IllegalArgumentException. Instead, we should scale the cursor image to match the current level.

Instead of reading the zoom level only once, it is now calculated dynamically, as the zoom may change depending on the display or because it has been changed. The methods for handling this case have also been moved from the SharedCursors class to the InternalGEFPlugin, so that it becomes available to the DragCursors class.

From there, we call asImageDataProvider() to convert the image descriptor to image data which internally scales the image via DPIUtil magic, if necessary.

See https://github.com/eclipse/gef-classic/issues/464

### CODE REVIEW

**Comment by ptziegler** [comment]
I've created https://github.com/eclipse-platform/eclipse.platform.ui/pull/2084 to contribute 2x variants for those drag icons. Everything seems to work for 100% zoom already, but I haven't had the chance to test this on a high-DPI screen.

File: org.eclipse.gef/src/org/eclipse/gef/internal/InternalGEFPlugin.java [hunk 1]
```diff
@@ -58,6 +60,26 @@ public static AbstractUIPlugin getDefault() {
 		return singleton;
 	}
 
+	public static int getDeviceZoom() {
```
Comment (fedejeanne) [inline org.eclipse.gef/src/org/eclipse/gef/internal/InternalGEFPlugin.java:63]: This implementation will probably suffer from the same problem that `DPIUtil` was suffering which is: the value will be obtained once and never updated again _i.e._ if the zoom of the monitor changes then this value will be outdated.

Comment (ptziegler) [reply org.eclipse.gef/src/org/eclipse/gef/internal/InternalGEFPlugin.java:63]: How was this solved in SWT?

Comment (fedejeanne) [reply org.eclipse.gef/src/org/eclipse/gef/internal/InternalGEFPlugin.java:63]: `DPIUtil.setDeviceZoom` is called on zoom changes, which keeps the `deviceZoom` always up-to-date _i.e._ calling `DPIUtil.getDeviceZoom()` should always give you the current zoom. But those methods are not accessible in this case, so this is just fyi.

In this specific case, you could simply avoid storing the zoom in the field `deviceZoom` and query the system property `org.eclipse.swt.internal.deviceZoom` every time. Since `DPIUtil.setDeviceZoom(...)` is **also** updating that property then the result should be that querying it always gives you the right zoom :-)

Long story short: remove the field `deviceZoom` and remove the condition `if (deviceZoom == -1)`.

**Comment by Phillipus** [comment]
If PNG images are used for the cursors then separate source and mask are not needed as the PNG format has transparency.

**Comment by ptziegler** [comment]
> If PNG images are used for the cursors then separate source and mask are not needed as the PNG format has transparency.

The new images still have a white background, to make sure that they look exactly the same.

File: org.eclipse.gef/src/org/eclipse/gef/internal/InternalGEFPlugin.java (hunk 1 above)
Comment (ptziegler) [reply org.eclipse.gef/src/org/eclipse/gef/internal/InternalGEFPlugin.java:63]: That you for the deeper insight in how this is handled on the SWT side of things. This is pretty similar to what I had in mind.

Though there is a separate issue which makes me concerned about how much we can get out of it in the first place. If the zoom changes, we normally would also need to refresh the cursors, as they are still using the old image data.

But those cursors are API, meaning we can't simply dispose them and create new objects...

Comment (ptziegler) [reply org.eclipse.gef/src/org/eclipse/gef/internal/InternalGEFPlugin.java]: With the latest state, the device zoom is no longer stored in a static variable, but always read directly from the system property.

**Comment by ptziegler** [comment]
At 100% zoom:
image

At 200% zoom:
image

At 300% zoom:
image

File: org.eclipse.gef/src/org/eclipse/gef/internal/InternalGEFPlugin.java (hunk 1 above)
Comment (fedejeanne) [reply org.eclipse.gef/src/org/eclipse/gef/internal/InternalGEFPlugin.java]: > That you for the deeper insight in how this is handled on the SWT side of things. This is pretty similar to what I had in mind.
> 
> Though there is a separate issue which makes me concerned about how much we can get out of it in the first place. If the zoom changes, we normally would also need to refresh the cursors, as they are still using the old image data.
> 
> But those cursors are API, meaning we can't simply dispose them and create new objects...

Yes, that is true. The problems are that:
1. The cursors are **constants** in `SharedCursors` _i.e._ one can not simply recreate them when the zoom changes
2. `Image` and `ImageData` don't really know anything about zoom changes which means once you create a `Cursor` out of an `Image`/`ImageData`, you're stuck with the zoom level at which they were created.

Point nr. 2 means that one has to manually react to zoom level changes and recreate the cursors with an image/image data at the appropriate zoom level, but this breaks the contract established in point nr. 1.

Sadly, I don't see an easy way out of this problem, at least not at this level of abstraction (GEF is a consumer of SWT). Best case scenario, this kind of issues would be handled transparently by SWT and not by the consumer (GEF), but I know this is not trivial since our team has been working on the [HighDPI subject](https://github.com/orgs/vi-eclipse/projects/1/views/1?filterQuery=label%3AHiDPI) for quite a while now and this issue has also surfaced [somewhere else](https://github.com/vi-eclipse/Eclipse-Platform/issues/65).

I will take this to the team and we'll try to address it in https://github.com/vi-eclipse/Eclipse-Platform/issues/65 too.

**Comment by azoitl** [comment]
> > If PNG images are used for the cursors then separate source and mask are not needed as the PNG format has transparency.
> 
> The new images still have a white background, to make sure that they look exactly the same.
> 
> Edit: Clients may also still run GEF with an older Eclipse version. By omitting the mask, we would break backwards compatibility.

@ptziegler we already have some cursors without mask since the last release. I hope I didn't break anything there. But the mask always made problems at least for these cursors. So I'm happy when it is gone.

**Comment by ptziegler** [comment]
@azoitl I talked to Heiko over at the Eclipse Platform PR and my current proposal is that for the new images, I drop the white background for transparency. In order to remain backwards compatible, we still need to use the mask. But for future cursors, it would then no longer be necessary.

File: org.eclipse.gef/src/org/eclipse/gef/internal/InternalGEFPlugin.java (hunk 1 above)
Comment (ptziegler) [reply org.eclipse.gef/src/org/eclipse/gef/internal/InternalGEFPlugin.java]: > Point nr. 2 means that one has to manually react to zoom level changes and recreate the cursors with an image/image data at the appropriate zoom level, but this breaks the contract established in point nr. 1.

That's what I'm afraid of. And it's not just affecting the SharedCursors class, but also the Draw2D Cursors class containing static instances of the native SWT cursors. In the long term, I believe the most sensible thing to do would be to deprecate those static fields and have something similar to the ImageDataProvider, except for cursors. 

GEF can then listen to SWT.ZoomChanged events and reload the cursors accordingly. But that is going way beyond the scope of this PR and something we have to consider at a later date. Thank you again for your input.

**Comment by azoitl** [comment]
> @azoitl I talked to Heiko over at the Eclipse Platform PR and my current proposal is that for the new images, I drop the white background for transparency. In order to remain backwards compatible, we still need to use the mask. But for future cursors, it would then no longer be necessary.

Thx for checking and the feedback.

**Comment by ptziegler** [comment]
As a quick update: With the 2024-09, the mask & source images are going to be marked as "for removal", being replaced by a new constants that combines both into one.

So instead of:

> createCursor(ISharedImages.IMG_OBJS_DND_INVALID_SOURCE, ISharedImages.IMG_OBJS_DND_INVALID_MASK);

One can simply do:

> createCursor(ISharedImages.IMG_OBJS_DND_INVALID);

**Comment by ptziegler** [comment]
So in principle, we could delay this PR until the M2 and then refactor the code or merge it now and keep the "old" images. The first would break backwards compatibility and require clients to also update to the latest Eclipse version, when updating GEF.

For my perspective, I would prefer to keep using the old icons and wait out the 2-year deprecation period.

**Comment by ptziegler** [comment]
I've updated the PR to also scale the shared images. While it can't handle dynamic DPI changes, it at least won't throw an IllegalArgumentException anymore.

File: org.eclipse.gef/src/org/eclipse/gef/internal/InternalGEFPlugin.java [hunk 2]
```diff
@@ -67,4 +72,45 @@ public void stop(BundleContext context) throws Exception {
 		super.stop(context);
 	}
 
+	/**
+	 * Convenience method for getting the current zoom level of the active device.If
+	 * on MacOS or Linux (x11 window system) or if the device zoom couldn't
+	 * otherwise be determined, this method returns {@code 100} as default value.
+	 */
+	@SuppressWarnings("nls")
```
Comment (fedejeanne) [inline org.eclipse.gef/src/org/eclipse/gef/internal/InternalGEFPlugin.java:80]: Better add the proper `//$NON-NLS-X$` and remove this, otherwise you'll be possibly silencing valid warnings that may be introduced in future changes.

File: org.eclipse.gef/src/org/eclipse/gef/internal/InternalGEFPlugin.java [hunk 3]
```diff
@@ -67,4 +72,45 @@ public void stop(BundleContext context) throws Exception {
...
+			return 100;
+		}
+	}
+
+	/**
+	 * Convenience class to get the image data for a given zoom level. If no image
+	 * for the requested zoom level exists, the image data may be an auto-scaled
+	 * version of the native image and may look more blurred or mangled than
+	 * expected.
+	 */
```
Comment (fedejeanne) [inline org.eclipse.gef/src/org/eclipse/gef/internal/InternalGEFPlugin.java:101]:
```suggestion
	/**
	 * Convenience method to get the image data for a given zoom level. If no image
	 * for the requested zoom level exists, the image data may be an auto-scaled
	 * version of the native image and may look blurred or mangled.
	 */
```

File: org.eclipse.gef/src/org/eclipse/gef/ui/palette/FlyoutPaletteComposite.java [hunk 4]
```diff
@@ -1557,37 +1559,42 @@ private static class DragCursors {
...
 				}
 			}
 			return cursors[code];
 		}
 
+		/**
+		 * Creates a new cursor using the shared images as source and mask,
+		 * respectively. The cursors are created with respect to the current device
+		 * zoom, with the hotspot being the center of the image.
+		 */
```
Comment (fedejeanne) [inline org.eclipse.gef/src/org/eclipse/gef/ui/palette/FlyoutPaletteComposite.java:1586]: Add the parameters to the javadoc and mention that they need to be names of shared cursors.

**Review by fedejeanne — COMMENTED** [review]
Just some nitpicks, the rest LGTM.

Regarding breaking backwards compatibility or not (https://github.com/eclipse/gef-classic/pull/479#issuecomment-2241762468) I'd say you make it clear that something will break and give time for the clients to adapt. If the appropriate methods are being deprecated _for removal_ anyway then there's no chance this change is going to be forgotten in 2 years anyway (one can live with the technical debt as long as there's an end in sight).

One last thing: I'd like to test this too. Can you please add a short "how to test" section to the description?

File: org.eclipse.gef/src/org/eclipse/gef/internal/InternalGEFPlugin.java (hunk 1 above)
Comment (ptziegler) [reply org.eclipse.gef/src/org/eclipse/gef/internal/InternalGEFPlugin.java]: > Which class is that? If it's using the native cursors (i.e. the constructor that takes a style as parameter) then they shouldn't be affected by this issue.

I'm talking about this [Cursors](https://github.com/eclipse/gef-classic/blob/master/org.eclipse.draw2d/src/org/eclipse/draw2d/Cursors.java) class, where we create static instances of the native cursors. Looking at the actual implementation, it looks like they don't actually work with ImageData instances, so you are indeed correct and they should be updated automatically.

File: org.eclipse.gef/src/org/eclipse/gef/internal/InternalGEFPlugin.java (hunk 2 above)
Comment (ptziegler) [reply org.eclipse.gef/src/org/eclipse/gef/internal/InternalGEFPlugin.java]: I've updated the PR to include the three things you've pointed out.

**Review by fedejeanne — APPROVED** [review]

**Review by azoitl — APPROVED** [review]
